src/lib/routes/admin-api/public-signup.ts

```diff
@@ -0,0 +1,252 @@
+import { CreateUserSchema } from '../../openapi/spec/create-user-schema';
+
+interface TokenParam {
+    token: string;
+}
+
+export class PublicSignupController extends Controller {
+    private publicSignupTokenService: PublicSignupTokenService;
+
+    private userService: UserService;
+
+    private accessService: AccessService;
+
+    private openApiService: OpenApiService;
+
+    private logger: Logger;
+
+    constructor(
+        config: IUnleashConfig,
+        {
+            publicSignupTokenService,
+            accessService,
+            userService,
+            openApiService,
+        }: Pick<
+            IUnleashServices,
+            | 'publicSignupTokenService'
+            | 'accessService'
+            | 'userService'
+            | 'openApiService'
+        >,
+    ) {
+        super(config);
+        this.publicSignupTokenService = publicSignupTokenService;
+        this.accessService = accessService;
+        this.userService = userService;
+        this.openApiService = openApiService;
+        this.logger = config.getLogger('public-signup-controller.js');
+
+        this.route({
+            method: 'get',
+            path: '/tokens',
+            handler: this.getAllPublicSignupTokens,
+            permission: ADMIN,
+            middleware: [
+                openApiService.validPath({
+                    tags: ['Public signup tokens'],
+                    operationId: 'getAllPublicSignupTokens',
+                    responses: {
+                        200: createResponseSchema('publicSignupTokenSchema'),
+                    },
+                }),
+            ],
+        });
+
+        this.route({
+            method: 'post',
+            path: '/tokens',
+            handler: this.createPublicSignupToken,
+            permission: ADMIN,
+            middleware: [
+                openApiService.validPath({
+                    tags: ['Public signup tokens'],
+                    operationId: 'createApiToken',
+                    requestBody: createRequestSchema(
+                        'publicSignupTokenCreateSchema',
+                    ),
+                    responses: {
+                        201: createResponseSchema('publicSignupTokenSchema'),
+                    },
+                }),
+            ],
+        });
+
+        this.route({
+            method: 'post',
+            path: '/tokens/:token/signup',
+            handler: this.addTokenUser,
+            permission: NONE,
+            middleware: [
+                openApiService.validPath({
+                    tags: ['Public signup tokens'],
+                    operationId: 'addPublicSignupTokenUser',
+                    requestBody: createRequestSchema('createUserSchema'),
+                    responses: {
+                        200: emptyResponse,
+                    },
+                }),
+            ],
+        });
+
+        this.route({
```

Comment:
Think we should add a separate endpoint for validating the token, and add ADMIN to this.